components/script/body.rs

```diff
@@ -91,13 +92,18 @@ impl TransmitBodyConnectHandler {
             bytes_sender: None,
             control_sender,
             in_memory,
+            in_memory_done: false,
             source,
         }
     }
 
     /// Re-extract the source to support streaming it again for a re-direct.
     /// TODO: actually re-extract the source, instead of just cloning data, to support Blob.
-    fn re_extract(&self, chunk_request_receiver: IpcReceiver<BodyChunkRequest>) {
+    fn re_extract(&mut self, chunk_request_receiver: IpcReceiver<BodyChunkRequest>) {
+        // Reset `in_memory_done` because this means there are
+        // incoming streams and this is not done yet.
+        self.in_memory_done = false;
```

Comment:
This is probably fine, since i don't think we will receive a chunk request following this, however it's strictly-speaking wrong, because what we want to do is leave `true` on the original, and only reset the clone to `false`. 

Maybe we need our own method that isn't `clone` to do this. We could also implement `Clone`, but I'm not sure if it's a good idea to have it include these side-effects.

Reply:
Do you mean doing like this 👀? I don't have confidence that this is a good idea either though, if I understand correctly, this should be same as "leave `true` on the original route, and only reset the clone to `false`" logic 🤔?

```rust
impl Clone for TransmitBodyConnectHandler {
    fn clone(&self) -> Self {
        Self {
            stream: self.stream.clone(),
            task_source: self.task_source.clone(),
            canceller: self.canceller.clone(),
            bytes_sender: self.bytes_sender.clone(),
            control_sender: self.control_sender.clone(),
            in_memory: self.in_memory.clone(),
            source: self.source.clone(),
            in_memory_done: false,
        }
    }
}
```

I would like to cc @jdm to see his point of view of doing clone like this 🙋‍♂️

Reply:
I think you can just have a method called `reset_in_memory_done`, that would set it to false, and be called on the clone after the cloning.

Reply:
hmm, I'm not sure if I get your point, because `RequestBody` only knows the IPC channel but doesn't know about the `in_memory_done` flag. If we'd like to add a method called `reset_in_memory_done`, does that mean we will also need a new IPC route (like `BodyChunkRequest::Reset`) to reset that flag? 🤔

Reply:
I think you can add the method to `TransmitBodyConnectHandler`, and then call it after it has been cloned.

Reply:
ahhh, I see! because we basically need to handle the reset for redirection and we will call `Extract` for the redirection, then we can call the function right after `re_extract`.

https://github.com/servo/servo/blob/e928b3865e03f9ab7cd19ef23b2cf0aaeca3707b/components/script/body.rs#L117-L119

is that correct? (because we do the clone above it)

Reply:
https://github.com/servo/servo/blob/e928b3865e03f9ab7cd19ef23b2cf0aaeca3707b/components/script/body.rs#L118 is for nested re-redirections, the initial call is at https://github.com/servo/servo/blob/e928b3865e03f9ab7cd19ef23b2cf0aaeca3707b/components/script/body.rs#L371

Reply:
I _might_ get your point now 🤔 so basically your point is, we should do the reset to the cloned body handler only instead of the original (a.k.a `self`) handler, is that correct?

Thanks for your patient >_<

Reply:
yes, that's it.